File: js/069_Sqrtx.js
/**
 * @param {number} x
 * @return {number}
 */
var mySqrt = function(x) {
    if( x === 0 ) return 0;
    var l = 1;
    var h = x;
    while(true){
        var m = l + Math.floor((h-l)/2);
        if( m === Math.floor(x/m) ) return m;
        else if( m < Math.floor(x/m)){
            if(m+1 > Math.floor( x/(m+1) )) return m;
            l = m + 1;
        }else{
            h = m - 1;
        }
    }
};

File: cpp/069_Sqrtx.cpp
class Solution {
public:
    int mySqrt(int x) {
        if( x == 0) return 0;
        int l = 1;
        int h = x;
        while(1){
            int m = l + (h-l)/2;
            if( m == x/m) return m;
            else if( m < x/m ){
                if(m+1 > x/(m+1)) return m;
                l = m+1;
            }else{
                h = m-1;
            }
        }
    }
};